File: generated_code/GeminiPro/Java/043_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree for an integer set.
 * A splay tree is a self-adjusting binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 */
public class SplayTree {

    private Node root;

    /**
     * Inner class representing a node in the splay tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Searches for a key in the splay tree.
     * This operation performs a splay, moving the accessed node (or the last
     * node on the search path) to the root of the tree.
     *
     * @param key The key to search for.
     * @return The key if it is found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the node with that key is splayed to the root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree on the key. This brings the node with the key,
        // or its would-be parent, to the root.
        root = splay(root, key);

        // If the key is already in the tree, we're done.
        if (root.key == key) {
            return;
        }

        // Otherwise, insert the new key at the root.
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not found, the tree is splayed on the last accessed node.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree on the key.
        root = splay(root, key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // The node to delete is now the root.
        // We join its left and right subtrees to form the new tree.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Splay the maximum element in the left subtree to its root.
            // Since `key` is greater than any element in the left subtree, splaying
            // for `key` will bring the maximum element to the top.
            Node newRoot = splay(leftSubtree, key);
            // Attach the original right subtree.
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }

    /**
     * Performs a right rotation on the subtree rooted at x.
     *
     * @param x The root of the subtree.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     *
     * @param x The root of the subtree.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splay operation, implemented recursively (bottom-up).
     * It moves the node with the given key (or the last accessed node on the
     * search path) to the root of the subtree.
     *
     * @param root The root of the subtree to splay.
     * @param key The key to splay on.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node root, int key) {
        // Base case: root is null or key is present at root
        if (root == null || root.key == key) {
            return root;
        }

        // Key lies in the left subtree
        if (key < root.key) {
            // Key is not in tree, we are done
            if (root.left == null) {
                return root;
            }

            // Zig-Zig (Left Left)
            if (key < root.left.key) {
                root.left.left = splay(root.left.left, key);
                root = rightRotate(root);
            }
            // Zig-Zag (Left Right)
            else if (key > root.left.key) {
                root.left.right = splay(root.left.right, key);
                if (root.left.right != null) {
                    root.left = leftRotate(root.left);
                }
            }
            
            // Do second rotation for root
            return (root.left == null) ? root : rightRotate(root);

        } else { // Key lies in the right subtree
            // Key is not in tree, we are done
            if (root.right == null) {
                return root;
            }

            // Zig-Zag (Right Left)
            if (key < root.right.key) {
                root.right.left = splay(root.right.left, key);
                if (root.right.left != null) {
                    root.right = rightRotate(root.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > root.right.key) {
                root.right.right = splay(root.right.right, key);
                root = leftRotate(root);
            }
            
            // Do second rotation for root
            return (root.right == null) ? root : leftRotate(root);
        }
    }
}

File: generated_code/GeminiPro/Python/043_SplayTree.py
class SplayTree:
    """
    A self-contained Splay Tree class implementing a dictionary-like set for integers.

    This implementation includes insert, delete, and search operations. The key
    feature of a splay tree is that recently accessed elements are moved to the
    root of the tree to optimize for temporal locality of reference, providing
    amortized O(log n) time complexity for its operations.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key, parent=None, left=None, right=None):
            self.key = key
            self.parent = parent
            self.left = left
            self.right = right

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            # x was the root, y is the new root
            # Note: self.root is not updated here; the caller is responsible.
            pass
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            # x was the root, y is the new root
            # Note: self.root is not updated here; the caller is responsible.
            pass
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.

        Returns the new root of the splayed tree (which is x).
        """
        if not x:
            return None
            
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig step
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:
                # Zig-Zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:
                # Zig-Zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:
                # Zig-Zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:
                # Zig-Zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)
        return x

    def search(self, key):
        """
        Searches for a key in the tree.

        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last accessed node (the would-be parent)
        is splayed to the root.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                # Key found, splay the node
                self.root = self._splay(node)
                return True

        # Key not found. If the tree was not empty, splay the last visited node.
        if last_node:
            self.root = self._splay(last_node)
            
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        # Case 1: Tree is empty
        if not self.root:
            self.root = self._Node(key)
            return

        # Case 2: Tree is not empty, find insertion point
        current = self.root
        parent = None
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay the found node and we are done.
                self.root = self._splay(current)
                return

        # Insert the new node
        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the newly inserted node
        self.root = self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        This is done by splaying the node to the root, removing it, and then
        joining the two resulting subtrees.

        Args:
            key (int): The integer key to delete.
        """
        # First, search for the key. This will splay the node if it exists,
        # or a nearby node if it doesn't.
        if not self.search(key):
            # Key is not in the tree, and search() has already splayed the
            # last accessed node. Nothing more to do.
            return

        # After search, if the key was found, it's now at the root.
        # self.root.key will be equal to key.
        
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left child, so the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # The left subtree exists. We will make its largest element the new root.
            # Detach the left subtree to work on it independently.
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree.
            new_root = left_subtree
            while new_root.right:
                new_root = new_root.right
            
            # Splay this maximum element to the root of the left subtree.
            # After this, new_root will be the root and have no right child.
            new_root = self._splay(new_root)

            # Join the original right subtree to the new root.
            new_root.right = right_subtree
            if right_subtree:
                right_subtree.parent = new_root
            
            self.root = new_root